File: AEDS/aeds2/tp3/Quicksort/Quicksort.java
/**
 * @path TP03Q010 - Lista Sequencial em Java
 * @author Rayssa Mell de Souza Silva
 */

//-------------- imports
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Scanner;

public class Quicksort {
    private int comparacoes;
    private int movimentacoes;

    //Construtor
    public Quicksort(){
        comparacoes = 0;
        movimentacoes = 0;
    }

    // Retorna o timestamp atual
	public long now(){
		return new Date().getTime();
	}

    // Gravar no arquivo tempo de execucao e numero de movimentacoes
    public void gravarArquivo (double inicio, double fim) {
        
        try (BufferedWriter buffWrite = new BufferedWriter(new FileWriter("matricula_quicksort3.txt"))) {
            double tempo = (fim - inicio) / 1000.0;  // Converte para segundos

            // Cria a string a ser gravada no arquivo
            String linha = "Matricula: 860210 \t" +
                           "Tempo de Execução: " + tempo + " s\t" +
                           "Número de Comparacoes: " + comparacoes + "\t" +
                           "Número de Movimentações: " + movimentacoes + "\t";

            // Escreve no arquivo
            buffWrite.write(linha);

        } catch (IOException e) {
            System.err.println("Erro ao gravar o arquivo: " + e.getMessage());
        }
    }

    //----------------- Metodo Principal
    public static void main(String[] args) throws Exception {
        Scanner sc = new Scanner(System.in);
        double inicio, fim;
        Quicksort quicksort = new Quicksort();

        GerenciadorPokemons gp = new GerenciadorPokemons();
        gp.ler(); // ler do arquivo

        ListaDupla lista = new ListaDupla(); 
        String line;
        Pokemon pk = new Pokemon(); 

        while(!(line = sc.next()).equals("FIM")){
            int id = Integer.parseInt(line); //ler id
            pk = gp.buscarPokemon(id); //buscar pokemon do id
            lista.inserirFim(pk); //inserir na lista
        }
        
        // Tempo de execucaoo
        inicio = quicksort.now();
        quicksort.quicksort(lista);
        fim = quicksort.now();
        quicksort.gravarArquivo(inicio, fim);

        lista.print(); //mostrar lista

        sc.close();
    }

    //Buscar celula do meio da lista encadeada
    CelulaDupla buscarMeio(CelulaDupla esq, CelulaDupla dir) {
        CelulaDupla cel = esq;
        int count = 0;
    
        // Contar o numero de celulas entre 'esq' e 'dir'
        while (cel != null && cel != dir) {
            count++;
            cel = cel.prox;
        }
    
        // encontrar a celula do meio
        cel = esq;
        for (int i = 0; i < count / 2; i++) {
            if (cel != null) {
                cel = cel.prox;
            }
        }
    
        return cel; // Retorna a celula
    }

    //---------- --------- Swap para ordenacao
    void swap(CelulaDupla i, CelulaDupla j) {
        Pokemon temp = i.pokemon;
        i.pokemon = j.pokemon;
        j.pokemon = temp;
    }

    //-------- Ordenacao por Quicksort - Generation ou Name
    private void quicksort(CelulaDupla esq, CelulaDupla dir) {
        if (esq == null || dir == null || esq == dir || esq == dir.prox) return;
    
        CelulaDupla i = esq;
        CelulaDupla j = dir;
    
        // Seleciona o pivo para geracao e nome
        CelulaDupla meio = buscarMeio(esq, dir);
        String pivoNome = meio.pokemon.getName(); 
        int pivo = meio.pokemon.getGeneration(); 
    
        while (i != null && j != null && i != j && i != j.prox) {
            // Comparacao por geração e, se igual, por nome
            while (i.pokemon.getGeneration() < pivo || 
                  (i.pokemon.getGeneration() == pivo && i.pokemon.getName().compareTo(pivoNome) < 0)) { 
                i = i.prox;
                comparacoes++;
            }
    
            while (j.pokemon.getGeneration() > pivo || 
                  (j.pokemon.getGeneration() == pivo && j.pokemon.getName().compareTo(pivoNome) > 0)) {
                j = j.ant;
                comparacoes++;
            }
    
            if (i != null && j != null && i != j && i != j.prox) {
                swap(i, j); //swap para troca
                movimentacoes += 3;  // Contar 3 movimentações para o swap
            }
        }
    
        // Chamadas recursivas para as partes esquerda e direita
        if (esq != j && j != null) quicksort(esq, j.ant);
        if (i != dir && i != null) quicksort(i.prox, dir);
    }

    //Chamada para recursividade
    public void quicksort(ListaDupla lista) {
        quicksort(lista.primeiro.prox, lista.ultimo);
    }
}

//------------------- Classe LISTA DUPLA ---------------------//
class ListaDupla {
    public CelulaDupla primeiro;
    public CelulaDupla ultimo;
    public int n; //tamanho

    //formatador de data
    SimpleDateFormat formatter = new SimpleDateFormat("dd/MM/yyyy");

    //Construtor Padrao
    public ListaDupla() {
		primeiro = new CelulaDupla();
		ultimo = primeiro;
	}

    public void inserirFim(Pokemon pk) {
		ultimo.prox = new CelulaDupla(pk);
        ultimo.prox.ant = ultimo;
		ultimo = ultimo.prox;
        n++;
	}

    //---------------------- Metodo Imprimir
    public void print(){
        for(CelulaDupla i=primeiro.prox; i !=null; i=i.prox){
            System.out.println("[#" 
                            + i.pokemon.getId() + " -> " 
                            + i.pokemon.getName() + ": " 
                            + i.pokemon.getDescription() + " - " 
                            + i.pokemon.getTypes() + " - " 
                            + i.pokemon.getAbilities()  + " - " 
                            + i.pokemon.getWeight() + "kg - " 
                            + i.pokemon.getHeight() + "m - "  
                            + i.pokemon.getCaptureRate() + "% - "  
                            + i.pokemon.getIsLegendary() + " - "  
                            + i.pokemon.getGeneration() + " gen] - "  
                            + (formatter.format(i.pokemon.getCaptureDate())));
        }
    }
}

//------------------- Classe Celula ---------------------//
class CelulaDupla {
    public Pokemon pokemon;
    public CelulaDupla prox;
    public CelulaDupla ant;

    public CelulaDupla() {
		this(null);
	}

    public CelulaDupla(Pokemon pokemon){
        this.pokemon = pokemon;
        this.prox = null;
        this.ant = null;
    }
}

//------------------- Classe POKEMON ---------------------//
class Pokemon {
    //atributos
    private int id;
    private int generation;
    private String name;
    private String description;
    private ArrayList<String> types;
    private ArrayList<String> abilities;
    private double weight;
    private double height;
    private int captureRate;
    private boolean isLegendary;
    private Date captureDate;
    
    //Construtor com valores vazios
    public Pokemon(){
        id = 0;
        generation = 0;
        name = "";
        description = "";
        types = new ArrayList<>();
        abilities = new ArrayList<>();
        weight = 0.0;
        height = 0.0;
        captureRate = 0;
        isLegendary = false;
        captureDate = new Date();
    }

    //Construtor com atribuicao de valores
    public Pokemon(int id, int generation, String name, String description, ArrayList<String> types, ArrayList<String> abilities, 
                   double weight, double height, int captureRate, boolean isLegendary, Date captureDate){
        this.id = id;
        this.generation = generation;
        this.name = name;
        this.description = description;
        this.types = types;
        this.abilities = abilities;
        this.weight = weight;
        this.height = height;
        this.captureRate = captureRate;
        this.isLegendary = isLegendary;
        this.captureDate = captureDate;
    }

    //----------------- Metodos get e set
    public int getId(){
        return id;
    }
    public void setId(int id){
        this.id = id;
    }
    public int getGeneration(){
        return generation;
    }
    public void setGeneration(int generation){
        this.generation = generation;
    }
    public String getName(){
        return name;
    }
    public void setName(String name){
        this.name = name;
    }
    public String getDescription(){
        return description;
    }
    public void setDescription(String description){
        this.description = description;
    }
    public ArrayList<String> getTypes(){
        return types; 
    }
    public void setTypes(ArrayList<String> types){
        this.types = new ArrayList<>(types); 
    }
    public ArrayList<String> getAbilities(){
        return abilities;
    }
    public void setAbilities(ArrayList<String> abilities){
        this.abilities = new ArrayList<>(abilities); 
    }
    public double getWeight(){
        return weight;
    } 
    public void setWeight(double weight){
        this.weight = weight;
    }
    public double getHeight(){
        return height;
    }
    public void setHeight(double height){
        this.height = height;
    }
    public int getCaptureRate(){
        return captureRate;
    }
    public void setCaptureRate(int captureRate){
        this.captureRate = captureRate;
    }
    public boolean getIsLegendary(){
        return isLegendary;
    }
    public void setIsLegendary(boolean isLegendary){
        this.isLegendary = isLegendary;
    }
    public Date getCaptureDate(){
        return captureDate;
    }
    public void setCaptureDate(Date captureDate){
        this.captureDate = captureDate;
    }

    //---------------------- Metodo Clone
    @Override
    public Pokemon clone(){ 
        try {
            return (Pokemon) super.clone();
        } catch (CloneNotSupportedException e) {
            throw new AssertionError(e);
        }
    }
}

//------------------- Classe para Gerenciar Pokemons ------------//
class GerenciadorPokemons {
    //Lista para pokemons lidos do arquivo
    private List<Pokemon> pokemons = new ArrayList<>(801);

    //caminho para arquivo
    private static final String FILE_PATH = "/tmp/pokemon.csv";
    //formatador de data
    SimpleDateFormat formatter = new SimpleDateFormat("dd/MM/yyyy");

    //----------------------- Metodo para Ler Pokemons do arquivo e guardar na lista 
    public void ler() { 
        // leitura do arquivo
        try (BufferedReader br = new BufferedReader(new FileReader(FILE_PATH))) {
            
            String linha;
            int linhaCont = 0;

            while ((linha = br.readLine()) != null) {
                linhaCont++;
                // pula linha de atribuicoes
                if(linhaCont > 1){
                    // array de string com a linha separada por virgula e []
                    String[] partes = linha.split(",(?=(?:[^\"]*\"[^\"]*\")*[^\"]*$)");
                    Pokemon pokemon = new Pokemon();

                    // tratar dados 
                    for(int i = 1; i < partes.length; i++){
                        // se algum atributo for vazio coloca 0
                        if(partes[i].isEmpty()){
                            partes[i] = "0";
                        } 

                        pokemon.setId(Integer.parseInt(partes[0]));
                        pokemon.setGeneration(Integer.parseInt(partes[1]));
                        pokemon.setName(partes[2]);
                        pokemon.setDescription(partes[3]);

                        //---- TYPES -----//
                        ArrayList<String> types = new ArrayList<>();
                        String type1 = "'" + partes[4].trim() + "'";
                        types.add(type1);

                        // se o segundo tipo nao for vazio
                        if (!partes[5].equals("0")) {
                            String type2 = "'" + partes[5].trim()+ "'";
                            types.add(type2);
                        }
                        pokemon.setTypes(types);

                        // ----- Abilities -----//
                        ArrayList<String> abilities = new ArrayList<>();
                        String ability = "";

                        for(int j=0; j < partes[6].length(); j++){
                            char character = partes[6].charAt(j);
                            if (character != '\"' && character != '[' && character != ']') {
                                // Se encontrar uma virgula, a habilidade foi capturada
                                if (character == ',') {
                                    if (!ability.isEmpty()) {
                                        abilities.add(ability.trim());  // Adiciona a habilidade a lista
                                        ability = ""; // Limpa para capturar a próxima habilidade
                                    }
                                } else {
                                    ability += character;
                                }
                            }
                        }
                        if(!ability.equals("0")){
                            abilities.add(ability.trim()); 
                        }
                        pokemon.setAbilities(abilities);

                        //------ Adicionando demais dados ----///
                        pokemon.setWeight(partes[7].isEmpty() ? 0.0 : Double.parseDouble(partes[7]));
                        pokemon.setHeight(partes[8].isEmpty() ? 0.0 : Double.parseDouble(partes[8]));
                        pokemon.setCaptureRate(Integer.parseInt(partes[9]));
                        
                        int valorLido = Integer.parseInt(partes[10]);
                        pokemon.setIsLegendary(valorLido == 1);
                        
                        // tratar data
                        try {
                            Date data = formatter.parse(partes[11]);
                            pokemon.setCaptureDate(data);
                        } catch (ParseException e) {
                            e.printStackTrace(); 
                        }

                        // adiciona o pokemon a lista 
                        pokemons.add(pokemon);
                    }
                } 
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    //--------------------------- Metodo para inserir pokemon do ID indicado
    public Pokemon buscarPokemon(int id) {
        Pokemon pk = new Pokemon();
        for (int i = 0; i < pokemons.size(); i++) {
            Pokemon pok = pokemons.get(i);  // Obtem o Pokemon na posição i
            if (pok.getId() == id) {
                pk = pok;  // captura os dados do Pokémon encontrado
                i = pokemons.size(); // sai do loop
            }
        }
        return pk; // retorna pokemon encontrado
    }

    public List<Pokemon> getPokemons() {
        return pokemons;
    }
}

File: 2° semestre/aeds2/tp3/Quicksort/Quicksort.c
/**
 * @path TP03Q08 - Quicksort com Lista Duplamente Encadeada
 * @author Rayssa Mell de Souza Silva
 */

//-----------imports
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <time.h>

//caminho para arquivo
#define FILE_PATH "/tmp/pokemon.csv"

#define MAXTAM 801
#define MAX_ABILITIES 8

//---------- Struct Data
typedef struct Date {
    int dia;
    int mes;
    int ano;
} Date;

//----------- Struct Pokemon 
typedef struct {
    int id;
    int generation;
    char name[100];
    char description[100];
    char **types;
    char **abilities;
    float weight;
    float height;
    int captureRate;
    bool isLegendary;
    Date captureDate;
} Pokemon;

Pokemon *pokemon[MAXTAM]; 

int movimentacoes = 0; //contador de movimentacoes
int comparacoes = 0; // contador de comparacoes

//----------- Funcao para criar e inicializar um Pokemon
Pokemon* createPokemon() {
    Pokemon* newPokemon = (Pokemon*)malloc(sizeof(Pokemon));
    if (newPokemon == NULL) {
        printf("Erro ao alocar memoria para um novo Pokemon\n");
    }

    // Inicializando os campos
    newPokemon->id = -1;
    newPokemon->generation = -1;
    newPokemon->types = malloc(2 * sizeof(char*)); // Para dois tipos
    newPokemon->types[0] = NULL; // Inicializa como NULL
    newPokemon->types[1] = NULL; // Inicializa como NULL

    newPokemon->abilities = malloc(MAX_ABILITIES * sizeof(char*)); 
    newPokemon->abilities[0] = NULL; // Inicializa como NULL

    // Inicializa outros campos
    newPokemon->weight = 0.0;
    newPokemon->height = 0.0;
    newPokemon->captureRate = 0;
    newPokemon->isLegendary = false;
    newPokemon->generation = 0;
    newPokemon->captureDate.dia = 0;
    newPokemon->captureDate.mes = 0;
    newPokemon->captureDate.ano = 0;

    return newPokemon;
}

//----------------- Funcao para Clonar Pokemon
Pokemon* clone(Pokemon* original) {
    Pokemon* novoPokemon = (Pokemon*)malloc(sizeof(Pokemon));
    
    if (novoPokemon != NULL) {

        novoPokemon->id = original->id;
        novoPokemon->generation = original->generation;  
        novoPokemon->weight = original->weight;
        novoPokemon->height = original->height;
        novoPokemon->captureRate = original->captureRate;
        novoPokemon->isLegendary = original->isLegendary;
        novoPokemon->captureDate = original->captureDate;

        strcpy(novoPokemon->name, original->name);
        strcpy(novoPokemon->description, original->description);

        // Types
        novoPokemon->types = malloc(2 * sizeof(char*)); 
        for (int i = 0; i < 2; i++) {
            if (original->types[i] != NULL) {
                novoPokemon->types[i] = strdup(original->types[i]);
            } else {
                novoPokemon->types[i] = NULL;
            }
        }

        // Abilities
        novoPokemon->abilities = malloc(MAX_ABILITIES * sizeof(char*));
        for (int i = 0; i < MAX_ABILITIES; i++) {
            if (original->abilities[i] != NULL) {
                novoPokemon->abilities[i] = strdup(original->abilities[i]);
            } else {
                novoPokemon->abilities[i] = NULL;
            }
        }
    }
    return novoPokemon; // Retorna o Pokemon clonado
}

//----------------- Metodos get e set
int getId(Pokemon *pokemon){
    return pokemon->id;
}
void setId(Pokemon *pokemon, int id){
    pokemon->id = id;
}
int getGeneration(Pokemon *pokemon){
    return pokemon->generation;
}
void setGeneration(Pokemon *pokemon, int generation){
    pokemon->generation = generation;
}
char *getName(Pokemon *pokemon) {
    return pokemon->name;
}
void setName(Pokemon *pokemon, char *name) {
    strcpy(pokemon->name, name);
}
char *getDescription(Pokemon *pokemon){
    return pokemon->description;
}
char setDescription(Pokemon *pokemon, char *description){
    strcpy(pokemon->description, description);
}
char **getTypes(Pokemon *pokemon){
    return pokemon->types; 
}
void setTypes(Pokemon *pokemon, char **types){
    pokemon->types = types;
}
char **getAbilities(Pokemon *pokemon){
    return pokemon->abilities;
}
void setAbilities(Pokemon *pokemon, char **abilities){
    pokemon->abilities = abilities;
}
float getWeight(Pokemon *pokemon){
    return pokemon->weight;
} 
void setWeight(Pokemon *pokemon, float weight){
    pokemon->weight = weight;
}
float getHeight(Pokemon *pokemon){
    return pokemon->height;
}
void setHeight(Pokemon *pokemon, float height){
    pokemon->height = height;
}
int getCaptureRate(Pokemon *pokemon){
    return pokemon->captureRate;
}
void setCaptureRate(Pokemon *pokemon, int captureRate){
    pokemon->captureRate = captureRate;
}
bool getIsLegendary(Pokemon *pokemon){
    return pokemon->isLegendary;
}
void setIsLegendary(Pokemon *pokemon, bool isLegendary){
    pokemon->isLegendary =  isLegendary;
}

Date getCaptureDate(Pokemon *pokemon) {
    return pokemon->captureDate;
}

void setCaptureDate(Pokemon *pokemon, Date captureDate) {
    pokemon->captureDate = captureDate;
}

//--- Procedimento para tratar habilidades ---//
void tratarHabilidades(char* campo, Pokemon* pokemon) {
    int pos = 0;

    // Substitui aspas simples e colchetes por espaco
    for (int j = 0; campo[j] != '\0'; j++) {
        if (campo[j] == '\'' || campo[j] == '[' || campo[j] == ']') {
            campo[j] = ' ';
        }
    }
    char* habilidade = strtok(campo, ",");
    
    while (habilidade != NULL && pos < MAX_ABILITIES) {
        //Remover espaco do inicio e final da habilidade
        while (*habilidade == ' ') habilidade++; 
        char* fim = habilidade + strlen(habilidade) - 1;
        while (*fim == ' ') {
            fim--;
        }
        fim[1] = '\0'; 

        if (strlen(habilidade) > 0) {
            //aloca espaco para cada habilidade
            pokemon->abilities[pos] = (char *)malloc((strlen(habilidade) + 1) * sizeof(char));
            strcpy(pokemon->abilities[pos], habilidade); // Copia a habilidade
            pos++;
        }
        habilidade = strtok(NULL, ","); 
    }
    pokemon->abilities[pos] = NULL;
}

//--- Funcao para verificar se os campos estao vazios ---//
bool verificarCamposVazios(char* campo) {
    bool resp = false;
    if (strncmp(campo, ",,,", 3) == 0) {
        resp = true;
    }
    return resp;
}

//--------------- Metodo para ler pokemons de arquivo
void ler(){
    FILE *arquivo = fopen(FILE_PATH, "rt");
    char linha[803];
    char *campo;
    int i = 0;
    int linhaAtual = 0;
   
    if (arquivo != NULL) {
        while (fgets(linha, sizeof(linha), arquivo) != NULL && i < 802 ) {
            linhaAtual++;
            if (linhaAtual != 1) { //pula a primeira linha
                
                char linhaCopia[sizeof(linha)];
                strcpy(linhaCopia, linha);

                pokemon[i] = createPokemon();
                
                //--- ID ---//
                campo = strtok(linha, ",");
                setId(pokemon[i], atoi(campo));

                //--- Generation ---//
                campo = strtok(NULL, ",");
                setGeneration(pokemon[i], atoi(campo));

                //--- Name ---// 
                campo = strtok(NULL, ",");
                setName(pokemon[i], campo);
                
                //--- Description ---//
                campo = strtok(NULL, ",");
                setDescription(pokemon[i], campo);
              
                //--- Types ---//
                campo = strtok(NULL, ",");
                pokemon[i]->types[0] = strdup(campo);

                campo = strtok(NULL, "\"");
                if (campo && strcmp(campo, ",") != 0) {
                    int length = strlen(campo);
                    campo[length-1] = '\0';
                    pokemon[i]->types[1] = strdup(campo);
                } else {
                    pokemon[i]->types[1] = NULL; // Se nqo houver segundo tipo
                }

                 //--- Abilities ---//
                campo = strtok(NULL, "\"");
                if (campo) {
                    tratarHabilidades(campo, pokemon[i]);
                }
               
                //--- Tratando os Campos apos habilidades ---//
                char *posicao = strchr(linhaCopia, ']');
                if (posicao != NULL) {
                    posicao += 2; // pula ]"
                    
                    float weight = 0.0, height = 0.0;
                    int captureRate = 0, isLegendary = 0;
                    Date date = {0, 0, 0};

                    // Verificar se os campos de peso e altura estao vazios
                    if (verificarCamposVazios(posicao)) {
                        // Tenta ler os demais valores (taxa de captura, lendario, e data de captura)
                        int numValoresLidos = sscanf(posicao + 3, "%d,%d,%d/%d/%d", 
                                                        &captureRate, &isLegendary, 
                                                        &date.dia, &date.mes, &date.ano);

                        if (numValoresLidos == 5) { 
                            setCaptureRate(pokemon[i], captureRate);
                            setIsLegendary(pokemon[i], (isLegendary == 1));
                            setCaptureDate(pokemon[i], date);
                        } 
                    } else {
                        // Tenta ler todos os campos com sscanf
                        int numValoresLidos = sscanf(posicao, ",%f,%f,%d,%d,%d/%d/%d", 
                                                    &weight, &height, 
                                                    &captureRate, &isLegendary, 
                                                    &date.dia, &date.mes, &date.ano);

                        if (numValoresLidos == 7) { //peso, altura, captura, lendario, dia, mês, ano)
                            // Atribuir os valores capturados
                            setWeight(pokemon[i], weight);
                            setHeight(pokemon[i], height);
                            setCaptureRate(pokemon[i], captureRate);
                            setIsLegendary(pokemon[i], (isLegendary == 1));
                            setCaptureDate(pokemon[i], date);
                        }                                                   
                    }  
                }    
            }
            i++;
        }
        fclose(arquivo);
        
    } else {
        printf("Erro ao abrir o arquivo.\n");
    }
}

//------ Funcao para buscar pokemon pelo ID
Pokemon *buscarPokemon(int id){
    Pokemon *pk = NULL;

    for (int i = 0; i < MAXTAM; i++) {
        Pokemon *pok = pokemon[i];  // Obtem o Pokemon na posicao i
        if (pok != NULL && getId(pok) == id) {
            pk = pok;
            i = MAXTAM;
        }
    }
    return pk; //retorna o pokemon do ID
}

//-------------- LISTA DUPLA -------------------//

typedef struct CelulaDupla {
	Pokemon *elemento;        
	struct CelulaDupla* prox; 
   struct CelulaDupla* ant;  
} CelulaDupla;

typedef struct ListaDupla {
    CelulaDupla* primeiro;
    CelulaDupla* ultimo;
    int n; //tamanho
} ListaDupla;

CelulaDupla* novaCelulaDupla(Pokemon *elemento) {
   CelulaDupla* nova = (CelulaDupla*) malloc(sizeof(CelulaDupla));
   nova->elemento = elemento;
   nova->ant = nova->prox = NULL;
   return nova;
}

ListaDupla *novaListaDupla(){
    ListaDupla* listaDupla = (ListaDupla*) malloc(sizeof(ListaDupla));
    listaDupla->primeiro = novaCelulaDupla(NULL);
    listaDupla->ultimo = listaDupla->primeiro;
    listaDupla->n = 0;

    return listaDupla;
}

//Funcao para inserir no fim da lista
void inserirFim(Pokemon *pk, ListaDupla *listaDupla) {
   listaDupla->ultimo->prox = novaCelulaDupla(pk);
   listaDupla->ultimo->prox->ant = listaDupla->ultimo;
   listaDupla->ultimo = listaDupla->ultimo->prox;
   listaDupla->n++;
}

//Buscar celula do meio
CelulaDupla* buscarMeio(CelulaDupla* esq, CelulaDupla* dir) {
    CelulaDupla *cel;
    if (!esq || !dir) cel = NULL;  // Verificar se 'esq' e 'dir' são validos

    cel = esq;
    int count = 0;

    // Contar o numero de celulas entre 'esq' e 'dir'
    while (cel != NULL && cel != dir) {
        count++;
        cel = cel->prox;
    }

    // encontrar a celula do meio
    cel = esq;
    for (int i = 0; i < count / 2; i++) {
        cel = cel->prox;
    }

    return cel; // Retorna a celula
}

//---------- --------- Swap para ordenacao
void swap(CelulaDupla *i, CelulaDupla *j) {
    Pokemon *temp = i->elemento;
    i->elemento = j->elemento;
    j->elemento = temp;
}

// Ordenacao Quicksort - Recursiva
void quicksortRec(CelulaDupla* esq, CelulaDupla* dir) {

   if (esq == NULL || dir == NULL || esq == dir || esq == dir->prox) return;

    // Buscar o pivo - elemento do meio
    CelulaDupla* meio = buscarMeio(esq, dir);
    char *pivoNome = getName(meio->elemento);
    int pivo = getGeneration(meio->elemento);

    CelulaDupla* i = esq;
    CelulaDupla* j = dir;

    while (i != NULL && j != NULL && i != j && i != j->prox) {
        // Primeiro faz a comparacao pela geracao, se for igual compara o nome
        while (getGeneration(i->elemento) < pivo || 
            (getGeneration(i->elemento) == pivo && strcmp(getName(i->elemento), pivoNome) < 0)) {
            i = i->prox;
            comparacoes++;
        }

        while (getGeneration(j->elemento) > pivo || 
            (getGeneration(j->elemento) == pivo && strcmp(getName(j->elemento), pivoNome) > 0)) {
            j = j->ant;
            comparacoes++;
        }


        if (i != NULL && j != NULL && i != j && i != j->prox) {
            swap(i, j);
            movimentacoes += 3;  // Contar 3 movimentacoes
        }
    }

    if (esq != j) quicksortRec(esq, j->ant); 
    if (i != dir) quicksortRec(i->prox, dir);
}

// chamada recursiva
void quicksort(ListaDupla *listaDupla) {
    quicksortRec(listaDupla->primeiro->prox, listaDupla->ultimo);
}

//-------------- Metodo para imprimir Pokemons
void print(ListaDupla *listaDupla) {
    for(CelulaDupla *i=listaDupla->primeiro->prox; i != NULL; i=i->prox){
        Pokemon* pok = i->elemento;
        printf("[#%d -> %s: %s - ['%s'", pok->id, pok->name, pok->description, pok->types ? pok->types[0] : "N/A");
        
        if (pok->types[1] != NULL) {
            printf(", '%s'", pok->types[1]);
        }

        // Habilidades
        for (int j = 0; pok->abilities[j] != NULL; j++) {
            if (j == 0) {
                printf("] - [");
            }
            printf("'%s'", pok->abilities[j]);
            if (pok->abilities[j + 1] != NULL) {
                printf(", ");
            }
        }

        printf("] - %.1fkg - %.1fm - %d%% - %s - %d gen] - %02d/%02d/%d\n", 
            pok->weight,
            pok->height,
            pok->captureRate,
            pok->isLegendary ? "true" : "false", 
            pok->generation, 
            pok->captureDate.dia, 
            pok->captureDate.mes, 
            pok->captureDate.ano);
    }
}

//Arquivo com tempo de execucao, n° de comparacoes e movimentacoes
void gravarArquivo(double inicio, double fim){
    FILE *arquivo = fopen("matricula_quicksort2.txt", "wt");;
    double total = ((fim - inicio) / (double)CLOCKS_PER_SEC); 

    fprintf(arquivo, "Matricula: 860210\t");  
    fprintf(arquivo, "Numero de Comparações: %d\t", comparacoes);	
    fprintf(arquivo, "Numero de Movimentações: %d\t", movimentacoes);
    fprintf(arquivo, "Tempo de Execução: %lf\t", total);

    fclose(arquivo);
}

//--------- Metodo Principal
int main(){
    char line[30];
    double inicio, fim;
    ListaDupla *listaDupla = novaListaDupla();
    Pokemon *pk;

    ler(); 
    scanf(" %[^\n]", line);  

    while(strcmp(line, "FIM") != 0) {
        int id = atoi(line);
        pk = buscarPokemon(id);
        inserirFim(pk, listaDupla);
        
        scanf(" %[^\n]", line);  
	}
   
    //inicializando variaveis globais para contagem
    movimentacoes = 0;
    comparacoes = 0;
    
    // Marcando tempo de execucao para quicksort //
    inicio += clock();  
    quicksort(listaDupla); 
    fim += clock();

    print(listaDupla);
    gravarArquivo(inicio, fim);
}